winpr/libwinpr/sspi/Kerberos/kerberos.c

```diff
@@ -659,117 +659,154 @@ static BOOL kerberos_mk_tgt_token(SecBuffer* buf, int msg_type, char* sname, cha
 	return ret;
 }
 
-static BOOL kerberos_rd_tgt_token(const sspi_gss_data* token, char** target, krb5_data* ticket)
+static size_t append(char* dst, const char* src, size_t dstSize)
 {
-	WinPrAsn1Decoder dec;
-	WinPrAsn1Decoder dec2;
-	BOOL error = 0;
-	WinPrAsn1_tagId tag = 0;
-	WinPrAsn1_INTEGER val = 0;
-	size_t len = 0;
-	wStream s;
-	char* buf = NULL;
-	char* str = NULL;
-
-	WINPR_ASSERT(token);
-
-	WinPrAsn1Decoder_InitMem(&dec, WINPR_ASN1_DER, (BYTE*)token->data, token->length);
+	char* ndst = strncpy(dst, src, dstSize);
+	if (ndst < dst)
+		return SIZE_MAX;
+	return ndst - dst;
+}
 
-	/* KERB-TGT-REQUEST (SEQUENCE) */
-	if (!WinPrAsn1DecReadSequence(&dec, &dec2))
+static BOOL kerberos_rd_tgt_req(WinPrAsn1Decoder* dec, char** target)
+{
+	BOOL rc = FALSE;
+	char* str = NULL;
+	if (!target)
```

Comment:
Check not needed this called only from this module, we control the callers.

Reply:
that is not quite correct.
the function is called with and without the parameter being `NULL`.
the decision is based on expectations (`if (context->state == KERBEROS_STATE_TGT_REQ && tok_id == TOK_ID_TGT_REQ)`) whereas the actual call is based on data read from the `ASN1` stream.

Reply:
correct I've been fooled by the diff view